### PR TITLE
Add a HLSNode

### PR DESCRIPTION
**Pull request by alanjrogers**
- Add a new HLSNode that uses hls.js to load video data into a HTMLVideoElement
- Fix several bugs that assumed Video and Audio Nodes were the only MediaNode subclasses (volume + playbackRate)
- Make sure the HLSNode passes down id, playbackRate duration etc. for easier debugging outside of VideoContext.
- HLSNode _unload method will stop the HLSNode from buffering if/when the node gets unloaded.

### CODE REVIEW

File: src/SourceNodes/hlsnode.ts [hunk 1]
```diff
@@ -0,0 +1,159 @@
+import RenderGraph from "../rendergraph";
+import MediaNode from "./medianode";
+import Hls from "hls.js";
+
+const TYPE = "HLSNode";
+
+// TODO(danareyes): when playing the video the load
+// of the node adds a lag, less buffer time seems to
+// imporve that we should figure out why
```
Comment (deinakings) [inline src/SourceNodes/hlsnode.ts]: I think this TODO I completed, the lag when loading future nodes was solved with https://github.com/alanjrogers/VideoContext/pull/1/files#diff-4a90f1bec989071e09ba32a760237357bace5012ba4e4f62f48d0a1f926303cfL899-L903

File: src/SourceNodes/hlsnode.ts [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+            this.hls.attachMedia(video);
+            this.hls.loadSource(this.src);
+            this._element = video;
+
+            this.loaded = true;
+            this._hlsLoading = true;
+            this.hls.startLevel = this.hls.levels.length - 1;
+
+            // add this for debugging video elements
+            let container = document.getElementById("video-debug-container");
```
Comment (deinakings) [inline src/SourceNodes/hlsnode.ts]: this debug code should be behind the `debug` flag too

Comment (alanjrogers) [reply src/SourceNodes/hlsnode.ts]: I'm going to remove this actually and have the app do it's own debugging.

File: src/SourceNodes/hlsnode.ts [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+            this._hls.attachMedia(video);
+            this._hls.loadSource(this._src);
+            this._element = video;
+
+            this._loaded = true;
+            this._hlsLoading = true;
+            this._hls.startLevel = this._hls.levels.length - 1;
+        }
+        if (!this._hlsLoading) {
+            // this.hls.attachMedia(this._element as HTMLMediaElement);
```
Comment (deinakings) [inline src/SourceNodes/hlsnode.ts]: can we remove this comment?

File: src/SourceNodes/hlsnode.ts [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+            this._hls.loadSource(this._src);
+            this._hlsLoading = true;
+        }
+        super._load();
+    }
+
+    _unload() {
+        if (this._hlsLoading) {
+            this._hls.stopLoad();
+            // this.hls.detachMedia();
```
Comment (deinakings) [inline src/SourceNodes/hlsnode.ts]: same question here

File: src/SourceNodes/hlsnode.ts [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+            video.id = this._id;
+            this._hls.attachMedia(video);
+            this._hls.loadSource(this._src);
+            this._element = video;
+
+            this._loaded = true;
+            this._hlsLoading = true;
+            this._hls.startLevel = this._hls.levels.length - 1;
+        }
+        if (!this._hlsLoading) {
```
Comment (deinakings) [inline src/SourceNodes/hlsnode.ts]: maybe better to have an `else if` here?

File: src/SourceNodes/sourcenode.ts [hunk 6]
```diff
@@ -79,6 +80,7 @@ abstract class SourceNode extends GraphNode {
         this._callbacks = [];
         this._renderPaused = false;
         this._displayName = TYPE;
+        this._id = src;
```
Comment (deinakings) [inline src/SourceNodes/sourcenode.ts]: this `_id` will not be unique.. we can have the same source in several nodes right?
also **src** is not alway `string` looks like

Comment (alanjrogers) [reply src/SourceNodes/sourcenode.ts]: Good point! I'll re-use the unique id approach here instead.

**Review by deinakings — APPROVED** [review]
Looks good!

**Pull request merged**